man/prlctl.8
.TH prlctl 8 "14 April 2011" "@PRODUCT_NAME_SHORT@"
.SH NAME
prlctl \- utility for managing @PRODUCT_NAME_SHORT@ servers and virtual environments residing on them.
.SH SYNOPSIS
prlctl \fBcreate\fR <\fIve_name\fR> [\fB-t,--ostemplate\fR <\fIname\fR>|\fB-o,--ostype\fR <\fIname\fR|\fIlist\fR>|\fB-d,--distribution\fR <\fIname\fR|\fIlist\fR>] [\fB--vmtype ct|vm\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--changesid\fR] [\fB--no-hdd\fR] [\fB--uuid\fR <\fIuuid\fR>] [\fBOPTIONS\fR]
.PP
prlctl \fBbackup\fR <\fIve_id\fR|\fIve_name\fR> [\fB-f,--full\fR] [\fB-i,--incremental\fR] [\fB-s,--storage\fR <\fBuser[[:passwd]@server[:port] [\fB--description\fR <\fIdesc\fR>]\fR>] [\fB--no-compression\fR] [\fB--no-tunnel\fR] [\fB--no-reversed-delta\fR] [\fB--backup-path\fR <\fIpath\fR>]
.PP
prlctl \fBbackup-list\fR [\fIve_id\fR|\fIve_name\fR] [\fB-f,--full\fR] [\fB--localvms\fR] [\fB--vmtype ct|vm|all\fR] [\fB-s,--storage\fR <\fBuser[[:passwd]@server[:port]\fR>] [\fB--backup-path\fR <\fIpath\fR>]
.PP
prlctl \fBbackup-delete\fR {\fB<ve_id>\fR | \fB-t,--tag\fR <\fIbackup_id\fR>} [\fB--keep-chain\fR] [\fB-s,--storage\fR <\fBuser[[:passwd]@server[:port]\fR>] [\fB--backup-path\fR <\fIpath\fR>]
.PP
prlctl \fBrestore\fR {\fB<ve_id>\fR | \fB-t,--tag\fR <\fIbackup_id\fR>}  [\fB-s,--storage\fR <\fBuser[[:passwd]@server[:port]\fR>] [\fB--backup-path\fR <\fIpath\fR>]
[\fB-n,--name\fR <\fBnew_name\fR>] [\fB--dst\fR <\fIpath\fR>] [\fB--no-tunnel\fR] [\fB--live\fR]
.PP
prlctl \fBcapture\fR <\fIve_id\fR|\fIve_name\fR> [\fB--file\fR <\fIpath\fR>]
.PP
prlctl \fBclone\fR <\fIve_id\fR|\fIve_name\fR> \fB--name\fR <\fInew_name\fR> [\fB--template\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--changesid\fR] [\fB--linked\fR] [\fB--detach-external-hdd\fR <\fIyes\fR|\fIno\fR>] [\fBOPTIONS\fR]
.PP
prlctl \fBdelete\fR <\fIve_id\fR|\fIve_name\fR> [\fB--force\fR]
.PP
prlctl \fBinstalltools\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBexec\fR <\fIve_id\fR|\fIve_name\fR> [\fB--without-shell\fR] <\fBcommand\fR>
.PP
prlctl \fBenter\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBconsole\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBlist\fR [\fB-a,--all\fR] [\fB-L\fR] [\fB-o,--output \fIfield\fR[,\fIfield\fR...]] 
[\fB-s,--sort \fR<\fIfield\fR|-\fIfield\fR>] [\fB-t,--template\fR] 
[\fB--vmtype \fR<\fIct\fR|\fIvm\fR|\fIall\fR>] [\fB-j,--json\fR]
.PP
prlctl \fBlist\fR -i,--info [\fB-f,--full\fR] [\fIve_id\fR|\fIve_name\fR]
[\fB-t,--template\fR] [\fB--vmtype ct|vm|all\fR] [\fB-j, --json\fR] 
.PP
prlctl \fBmigrate\fR <[\fBsrc_node/\fR]\fBID\fR> <\fBdst_node\fR[\fB/NAME\fR]> [\fB--dst\fR <\fIpath\fR>]
[\fB--changesid\fR] [\fB--clone\fR|\fB--remove-src\fR] [\fB--no-compression\fR] [\fB--no-tunnel\fR] [\fB--ssh\fR <\fIoptions\fR>]
.PP
prlctl \fBpause\fR <\fIve_id\fR|\fIve_name\fR> [\fBOPTIONS\fR]
.PP
prlctl \fBproblem-report\fR <\fIve_id\fR|\fIve_name\fR> <\fB-d,--dump\fR|\fB-s,--send\fR [\fB--proxy\fR [\fBuser[:password]@proxyhost[:port]]\fR] [\fB--no-proxy\fR]> [\fBOPTIONS\fR]
.PP
prlctl \fBreinstall\fR <\fIve_id\fR|\fIve_name\fR> [\fB--no-backup\fR] [\fB--resetpwdb\fR] [\fB--ostemplates\fR <\fIname\fR>]
.PP
prlctl \fBregister\fR <\fIpath\fR> [\fB--preserve-uuid\fR | \fB--uuid\fR <\fIUUID\fR>] [\fB--regenerate-src-uuid\fR] [\fB--force\fR] [\fBOPTIONS\fR]
.PP
prlctl \fBreset\fR <\fIve_id\fR|\fIve_name\fR> [\fBOPTIONS\fR]
.PP
prlctl \fBresume\fR <\fIve_id\fR|\fIve_name\fR> [\fBOPTIONS\fR]
.PP
prlctl \fBrestart\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBct2vm\fR <\fICT_id\fR|\fICT_name\fR>
.PP
prlctl \fBreset-uptime\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBset\fR <\fIve_id\fR|\fIve_name\fR> [\fB--cpus\fR <\fIn\fR>] [\fB--cpu-sockets\fR <\fIn\fR>] [\fB--memsize\fR <\fIn\fR>] [\fB--videosize\fR <\fIn\fR>]
.br
	[\fB--3d-accelerate\fR <\fIoff\fR|\fIhighest\fR|\fIdx9\fR>] [\fB--vertical-sync\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--cpuunits\fR <\fIn\fR>] [\fB--cpulimit\fR <\fIn\fR>] [\fB--cpumask\fR <{\fIn\fR[,\fIn\fR,\fIn1\fR-\fIn2\fR]|\fIall\fR}>] [\fB--nodemask\fR <{\fIn\fR[,\fIn\fR,\fIn1\fR-\fIn2\fR]|\fIall\fR}>]
.br
	[\fB--cpu-hotplug\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--ioprio\fR <\fIpriority\fR>] [\fB--iolimit\fR <\fIlimit\fR>] [\fB--iopslimit\fR <\fIlimit\fR>]
.br
	[\fB--mem-hotplug\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--memguarantee\fR <\fIauto\fR|\fIvalue\fR>]
.br
	[\fB--description\fR <\fIdesc\fR>] [\fB--autostart\fR <\fIon\fR|\fIoff\fR|\fIauto\fR>] [\fB--autostart-delay\fR <\fIn\fR>]
.br
	[\fB--autostop\fR <\fIstop\fR|\fIsuspend\fR>]
.br
	[\fB--autocompact\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--tools-autoupdate\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--on-crash\fR <\fIpause\fR|\fIrestart\fR>[:\fIno-report\fR]]
.br
	[\fB--vnc-mode\fR <\fIauto\fR|\fImanual\fR|\fIoff\fR>] [\fB--vnc-port\fR <\fIport\fR>] [{\fB--vnc-passwd\fR <\fIpasswd\fR | \fI-\fR> | \fB--vnc-nopasswd\fR}] [\fB--vnc-address\fR <\fIaddress\fR>]
.br
	[\fB--name\fR <\fIname\fR>] [\fB--rename-ext-disks\fR] [\fB--userpasswd\fR <\fIuser:passwd\fR> [\fB--host-admin\fR <\fIname\fR>]]
.br
	[\fB--netfilter\fR <\fIdisabled\fR|\fIstateless\fR|\fIstateful\fR|\fIfull\fR>]
.br
	[\fB--swappages\fR \fIpages\fR[:\fIpages\fR] \fB--swap\fR \fIbytes\fR] [\fB--quotaugidlimit\fR \fInum\fR]
.br
	[\fB--features\fR \fIname\fR:\fBon\fR|\fBoff\fR[\fB,\fR...]]
.br
	[\fB--device-add\fR <\fIhdd\fR|\fIcdrom\fR|\fInet\fR|\fIfdd\fR|\fIserial\fR|\fIusb\fR> | \fB--device-del\fR <\fIname\fR> [\fB--destroy-image\fR|\fB--destroy-image-force\fR|\fB--detach-only\fR] | \fB--device-set\fR <\fIname\fR>]
.br
	[\fB--device-connect\fR <\fIname\fR> | \fB--device-disconnect\fR <\fIname\fR>]
.br
	[\fB--device-bootorder\fR "<\fIname1 name2 ...\fR>"]
.br
	[\fB--offline-management\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--hostname\fR <\fIhostname\fR>] [\fB--nameserver\fR <\fIaddr\fR>] [\fB--searchdomain\fR <\fIaddr\fR>]
.br
	[\fB--rate\fR <\fIclass\fR:\fIKBits\fR>] [\fB--ratebound\fR <\fIyes\fR|\fIno\fR>]
.br
	[\fB--applyconfig\fR <\fIconf\fR>] [\fBOPTIONS\fR]
.br
	[\fB--nested-virt\fR <\fIon\fR|\fIoff\fR>]
.br
	[\fB--efi-boot\fR <\fIon\fR|\fIoff\fR>] [\fB--select-boot-device\fR <\fIon\fR|\fIoff\fR>] [\fB--external-boot-device\fR <\fIname\fR>] [\fB--update-nvram\fR>]
.br
	[\fB--ha-enable\fR <\fIyes\fR|\fIno\fR>] [\fB--ha-prio\fR <\fInumber\fR>]
.br
	[\fB--template\fR <\fIyes\fR|\fIno\fR>]
.br
	[\fB--backup-path\fR <\fIpath\fR>]
.br
	[\fB--backup-add\fR <\fIbackup_id\fR> [\fB--disk\fR <\fIdisk_name\fR>]
.br
	[\fB--backup-del\fR <\fIbackup_id\fR|\fIall\fR>]
.PP
prlctl \fBstart\fR <\fIve_id\fR|\fIve_name\fR> [\fB--repair\fR] [\fBOPTIONS\fR]
.PP
prlctl \fBstop\fR <\fIve_id\fR|\fIve_name\fR> [\fB--kill\fR] [\fBOPTIONS\fR]
.PP
prlctl \fBsnapshot\fR <\fIve_id\fR|\fIname\fR> [\fB-n\fR,\fB--name\fR <\fIname\fR>] [\fB-d\fR,\fB--description\fR <\fIdesc\fR>]
.PP
prlctl \fBsnapshot-list\fR <\fIve_id\fR|\fIname\fR> [{\fB-t\fR,\fB--tree\fR | \fB-i\fR,\fB--id\fR <\fIsnapid\fR>}]
.PP
prlctl \fBsnapshot-delete\fR <\fIve_id\fR|\fIname\fR> \fB-i\fR,\fB--id\fR <\fIsnapid\fR> [\fB-c\fR,\fB--children\fR]
.PP
prlctl \fBsnapshot-switch\fR <\fIve_id\fR|\fIname\fR> \fB-i\fR,\fB--id\fR <\fIsnapid\fR>
.PP
prlctl \fBsuspend\fR <\fIve_id\fR|\fIve_name\fR> [\fBOPTIONS\fR]
.PP
prlctl \fBunregister\fR <\fIve_id\fR|\fIve_name\fR> [\fBOPTIONS\fR]
.PP
prlctl \fBmount\fR <\fIve_id\fR|\fIve_name\fR> [{\fB-o\fR <\fIro\fR|\fIrw\fR> | \fB--info\fR}]
.PP
prlctl \fBumount\fR <\fIve_id\fR|\fIve_name\fR>
.PP
prlctl \fBmove\fR <\fIve_id\fR|\fIve_name\fR> \fB--dst\fR <\fIpath\fR>
.PP
prlctl \fBstatistics\fR {<\fIve_id\fR|\fIve_name\fR>|\fB-a\fR,\fB--all\fR} [\fB--filter\fR <\fIfilter\fR>] [\fB--loop\fR]

.SH DESCRIPTION
The \fBprlctl\fR utility is used to manage @PRODUCT_NAME_SHORT@ servers and virtual environments (VEs) residing on them.
A virtual environment can be referred to by its ID or name assigned to the VE during its creation.
.SH OPTIONS
.SS Flags
The following flags can be used with the majority of \fBprlctl\fR commands.
.IP "\fB-l,--login user[[:passwd]@server\fR" 4
Connect to the remote @PRODUCT_NAME_SHORT@ server using the IP address or hostname of \fBserver\fR and the specified credentials (i.e. the \fBuser\fR username and \fBpasswd\fR password). If no connection parameters are specified, \fBprlctl\fR assumes that the command is run on the local server. 
.IP "\fB-p,--read-passwd\fR <\fIfile\fR>" 4
Use the password from the \fBfile\fR file to log in to the remote @PRODUCT_NAME_SHORT@ server whose connection parameters are specified after the \fB--login\fR option.
.IP "\fB-v,--verbose <n>\fR" 4
Configure the \fBprlctl\fR logging level.
.IP "\fB--timeout <sec>\fR" 4
Specify a custom operation timeout in seconds. By default, timeouts for all operation are unlimited.
.SS Managing virtual environments
.IP "\fBcreate\fR <\fIve_name\fR> \fB-t,--ostemplate\fR <\fIname\fR> [\fB--vmtype ct|vm\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--uuid\fR <\fIuuid\fR>] [\fB--changesid\fR]" 4
Create the virtual environment with the name of \fB<ve_name>\fR on the basis of the specified template. You can get the list of available templates using the \fBprlctl list -t\fR command.

Use the \fB--changesid\fR option to assign the new Windows-based virtual machine a new Windows security identifier (SID). Note: @PRODUCT_NAME_SHORT@ Guest Tools must be installed in the virtual machine.

Use the \fB--uuid\fR option to manually specify the UUID to use.
.IP "\fBcreate\fR <\fIve_name\fR> [\fB-o,--ostype\fR <\fIname\fR|\fIlist\fR>] [\fB--vmtype ct|vm\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--uuid\fR <\fIuuid\fR>] [\fB--no-hdd\fR]" 4
Create the virtual environment with the name of \fB<ve_name>\fR and optimize it for use with the operating system (OS) family specified after the \fB--ostype\fR option, respectively. You can get the list of available os types using the \fBprlctl create  ve_name -o list\fR command.
.IP "\fBcreate\fR <\fIve_name\fR> [\fB-d,--distribution\fR <\fIname\fR|\fIlist\fR>] [\fB--vmtype ct|vm\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--uuid\fR <\fIuuid\fR>] [\fB--no-hdd\fR]" 4
Create the virtual environment with the name of \fB<ve_name>\fR and optimize it for use with the operating system (OS) family specified after the \fB--distribution\fR option, respectively. You can get the list of available distributions using the \fBprlctl create  ve_name -d list\fR command.

Use the \fB--dst\fR option to set the path to the directory on the @PRODUCT_NAME_SHORT@ server where the \fB<ve_uuid>\fR directory with files of the newly created virtual environment will be stored.

Use the \fB--no-hdd\fR option to create virtual environment without hard disk drives.

After the virtual environment has been successfully created, you should install the corresponding operating system inside it.
.IP "\fBclone\fR <\fIve_id\fR|\fIve_name\fR> \fB--name\fR <\fInew_name\fR> [\fB--template\fR] [\fB--dst\fR <\fIpath\fR>] [\fB--changesid\fR] [\fB--linked\fR] [\fB--detach-external-hdd\fR <\fIyes\fR|\fIno\fR>]" 4
Make a copy of a virtual environment and name it <\fInew_name\fR>. The difference between the original and the clone is that the clone is assigned a new MAC address.

Use the \fB--template\fR option to create a template of a virtual environment.

Use the \fB--dst\fR option to specify the path to the directory where the \fB<ve_uuid>\fR directory with files of the cloned virtual environment will be stored. If this option is omitted, the clone will be created in the default directory.

Use the \fB--changesid\fR option to assign the clone a new Windows security identifier (SID). Note: @PRODUCT_NAME_SHORT@ Guest Tools must be installed in the original Windows-based virtual machine.

Use the \fB--linked\fR option is used to create a linked clone of the virtual environment.

Use the \fB--detach-external-hdd\fR <\fIyes\fR|\fIno\fR> option to specify whether to keep or remove hard disks located outside of the original virtual environment. If you specify \fIyes\fR, outside hard disks will be removed from the resulting configuration. If you specify \fIno\fR, outside hard disks will remain in the resulting configuration. Note: In either case, outside hard disks will not be copied to the destination.
 
.IP "\fBdelete\fR <\fIve_id\fR|\fIve_name\fR>" 4
Remove the specified virtual environment from the @PRODUCT_NAME_SHORT@ server by deleting all VE-related files and directories.
You can use the \fB--force\fR option to forcibly stop the VE.
.IP "\fBinstalltools\fR <\fIve_id\fR|\fIve_name\fR>" 4
Connect the @PRODUCT_NAME_SHORT@ Guest Tools image to the VE using its CD/DVD-ROM device, so that the guest operating system is able to install or update the tools from it.
.IP "\fBexec\fR <\fIve_id\fR|\fIve_name\fR> <\fBcommand\fR>" 4
Execute the \fBcommand\fR command in the virtual environment. For virtual machines, it requires @PRODUCT_NAME_SHORT@ Guest Tools to be installed. Commands in Linux and Windows guests are run via bash -c "command" and cmd /c "command", respectively.
If the \fB--without-shell\fR option is specified, the command is run directly without bash or cmd shell.
.IP "\fBenter\fR <\fIve_id\fR|\fIve_name\fR>" 4
Log in to the virtual environment. For virtual machines, it requires @PRODUCT_NAME_SHORT@ Guest Tools to be installed.
.IP "\fBconsole\fR <\fIve_id\fR|\fIve_name\fR>" 4
Attach to the Container's console. To exit from the console, press "Esc" then "."
(Note: This sequence is only recognized after Enter).
Note that you can even attach to a console if a Container is not yet running.
.IP "\fBcapture\fR <\fIve_id\fR|\fIve_name\fR> [\fB--file\fR <\fIpath\fR>]" 4
Take a screenshot of a current virtual environment console in PNG format. If the \fB--file\fR option is specified, a screenshot is stored to the path \fIpath\fR; otherwise, it is dumped to stdout.
.IP "\fBpause\fR <\fIve_id\fR|\fIve_name\fR>" 4
Pause the specified virtual environment.
.IP "\fBproblem-report\fR <\fIve_id\fR|\fIve_name\fR> <\fB-d,--dump\fR|\fB-s,--send\fR [\fB--proxy\fR [\fBuser[:password]@proxyhost[:port]]\fR] [\fB--no-proxy\fR]> [\fBOPTIONS\fR]" 4
Generate a problem report. If the \fB-s,--send\fR option is specified, the report is sent to @PRODUCT_NAME_SHORT@ development team; otherwise, it is dumped to stdout.
.IP "\fBregister\fR <\fIpath\fR> [\fB--preserve-uuid\fR | \fB--uuid\fR <\fIUUID\fR>] [\fB--regenerate-src-uuid\fR] [\fB--force\fR]" 4
Register the virtual environment whose configuration file has the path of \fIpath\fR.
If the \fB--preserve-uuid\fR option is specified, the virtual environment ID will not changed.
If the \fB--uuid\fR option is specified, the provided \fIUUID\fR will be used for virtual environment ID,
otherwise, it will be regenerated.
If the \fB--regenerate-src-uuid\fR option is specified, the virtual environment source ID will be regenerated (SMBIOS product id will be changed as well).
If the \fB--force\fR option is specified, all validation checks will be skipped.
.IP "\fBreinstall\fR <\fIve_id\fR|\fIve_name\fR> [\fB--no-backup\fR] [\fB--resetpwdb\fR] [\fB--ostemplates\fR <\fIname\fR>]" 4
Recreates container's root virtual disk, installs all applications previously
installed from application templates, copies credentials from the old
container (unless \fB--resetpwdb\fR is specified), and renames the old root
directory '/' to '/old' (unless \fB--no-backup\fR is specified).
.IP "\fBreset\fR <\fIve_id\fR|\fIve_name\fR>" 4
Reset the specified virtual environment.
.IP "\fBstart\fR <\fIve_id\fR|\fIve_name\fR> [\fB--repair\fR]" 4 
Start the specified virtual environment. If the VE is a container and the \fB--repair\fR option is specified,
the container will start in the repair mode.
.IP "\fBrestart\fR <\fIve_id\fR|\fIve_name\fR>" 4
Restart the specified virtual environment.
.IP "\fBstop\fR <\fIve_id\fR|\fIve_name\fR> [\fB--kill\fR]" 4
Stop the specified virtual environment. You can use the \fB--kill\fR option to forcibly stop the VE.
.IP "\fBstatus\fR <\fIve_id\fR|\fIve_name\fR>" 4
Display the status of the specified virtual environment.
.IP "\fBunregister\fR <\fIve_id\fR|\fIve_name\fR>" 4
Unregister the specified virtual environment.
.IP "\fBsuspend\fR <\fIve_id\fR|\fIve_name\fR>" 4
Suspend the specified virtual environment.
.IP "\fBresume\fR <\fIve_id\fR|\fIve_name\fR>" 4
Resume the specified virtual environment.
.IP "\fBconvert\fR <\fICT_id\fR|\fICT_name\fR>" 4
Convert the specified container to VM.
.IP "\fBreset-uptime\fR <\fIve_id\fR|\fIve_name\fR>" 4
Resets the specified virtual environment uptime counter (counter start date/time also will be reset with this action).
.IP "\fBmount\fR <\fIve_id\fR|\fIve_name\fR> [{\fB-o\fR <\fIro\fR|\fIrw\fR> | \fB--info\fR}]" 4
Mounts the specified virtual environment.
.IP "\fBumount\fR <\fIve_id\fR|\fIve_name\fR>" 4
Unmounts the specified virtual environment.
.IP "\fBmove\fR <\fIve_id\fR|\fIve_name\fR> \fB--dst\fR <\fIpath\fR>" 4
Moves the directory with files of the specified virtual environment to a new location on the same server.
.TP
\fBContainer action scripts\fR
.RS
.PP
Action scripts can be used to perform actions on containers at various stages of container operation. The following commands can trigger action scripts: \fBstart\fR, \fBstop\fR, \fBrestart\fR, \fBmount\fR, and \fBumount\fR.
.PP
Two types of scripts are supported: global, triggered for each container on host, and per-container, triggered for specific containers. Custom action scripts can be created manually and must be assigned specific file names to be triggered.
.PP
Custom action scripts can be of two types:
.sp -1
.HP 2
\(bu Global, executed for all containers on host. Such scripts must have the prefix 'vps' (e.g., 'vps.mount') and need to be placed in '/etc/vz/conf/'.
.br
.sp -1
.HP 2
\(bu per-container, executed for specific containers. Such scripts must not have the prefix 'vps' and need to be placed in '/vz/private/<CT_UUID>/scripts/'.
.IP "\fBvps.mount\fR, \fBmount\fR" 4
Executed after a container is mounted. Can be global or container-specific.
.IP "\fBstart\fR" 4
Executed in container context on container start.
.IP "\fBstop\fR" 4
Executed in container context on container stop.
.IP "\fBvps.umount\fR, \fBumount\fR" 4
Executed before a container unmounted. Can be global or container-specific.
.PP
\fBNOTE\fR: All action scripts except \fBstart\fR and \fBstop\fR are executed in the host context. The \fBstart\fR and \fBstop\fR scripts are executed in the container context.
.PP

bash_completion/prlctl.sh
get_vms() {
	# TODO if remote parameters are specified -
	# use them here
	prlctl list -a -Ho name 2>/dev/null \
		| while read name; do printf "%q\n" "$name"; done
}

get_uuids() {
	prlctl list -a -Ho uuid 2>/dev/null | tr -d {}
}

get_vm_ids() {
	printf "$(get_vms)\n$(get_uuids)\n"
}

get_snaps_names() {
	# TODO Don't know how to find snaps names easily
	true
}

get_snaps_ids() {
	# TODO Add checking of vm existence
	# TODO Don't work because of '{' :-(
	prlctl snapshot-list $1 2>/dev/null | awk '{if (NR > 1) print substr($0, 41, 38)}'
}

get_backups_ids() {
	prlctl backup-list $1 2>/dev/null | awk '{if (NR > 1) print $2}' | tr -d '{}'
}

get_ostypes() {
	# TODO How to get supported OS types?
	true
}

get_distros() {
	# TODO How to get supported distros?
	# TODO: "prlctl create ads --distribution dsa" shows available distros
	true
}

get_ostemplates() {
	# available templates
	prlctl list -t -o name 2>/dev/null | awk '{if (NR > 1) print $1}'
}

get_devs() {
	# TODO Add checking of vm existence and device type/mode
	vmid=$1
	devtype=$2
	prlctl list -i $vmid 2>/dev/null | egrep '^[[:space:]]+[[:alnum:]]+ \([+-]\)' | awk '{print $1}'
}

_prlctl()
{
	local cur prev opts base
	COMPREPELY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"

	local actions_on_vmid="capture clone delete installtools enter migrate \
pause reset resume start stop snapshot snapshot-delete \
snapshot-list snapshot-switch suspend statistics unregister \
set backup restore backup-list backup-delete reset-uptime \
move exec console mount umount status problem-report change-sid \
restart list ct2vm"
	local actions_without_vmid="create list register server"

	local capture_flags='--file'
	local clone_flags='--name'
	local clone_optional_flags='--template --location'
	local create_flags='-c --config --location -o --ostype -d --distribution --ostemplate'
	local list_flags='-a --all -t --template -o --output -s --sort -i --info -f --full -j --json --vmtype'
	local migrate_flags='--location --no-compression'
	local snapshot_flags='-n --name -d --description'
	local snapshotlist_flags='-t --tree -i --id'
	local snapshotdelete_flags='-i --id'
	local snapshotswitch_flags='-i --id'
	local backup_flags='--description -f --full -i --incremental --no-compression'
	local backuplist_flags='-f --full --localvms'
	local backupdelete_flags='-t --tag'
	local restore_flags='-t --tag'
	local statistics_flags='--loop --filter'
	local set_flags='--cpus --memsize --videosize --description \
--onboot --name --device-add --device-del --device-set \
--device-connect --device-disconnect --applyconfig \
--netfilter --swappages --swap --quotaugidlimit \
--autostart --autostart-delay --autostop \
--vnc-mode --vnc-port --vnc-passwd --ioprio \
--ha-enable --ha-prio \
--device-bootorder --efi-boot --select-boot-device \
--external-boot-device --external-boot-device --update-nvram'
	local set_device_hdd_flags='--image --type --size --split --iface --position --device --passthr \
--encrypt --decrypt --encryption-keyid --reencrypt --nowipe'
	local set_device_cdrom_flags='--device --image --iface --position --passthr'
	local set_device_net_flags='--type --mac --iface'
	local set_device_fdd_flags='--device --image --recreate'
	local set_device_serial_flags='--device --output --socket'
	local move_flags='--location'

	local global_flags='-l -p -v --verbose'

	if [ $COMP_CWORD == 1 ]; then
		opts="${actions_on_vmid// /$'\n'}\n${actions_without_vmid// /$'\n'}"
	else
		for i in $actions_on_vmid; do
			if [ "${prev}" = "${i}" ]; then
				if [ -z "${cur}" ]; then
					opts=$(get_vms)
				else
					opts=$(get_vm_ids)
				fi
				break
			fi
		done

	fi

	if [ -z "${opts}" ]; then

		# processing all options that require arguments
		case "${prev}" in
		-n|--name)
			if [ "${COMP_WORDS[1]}" = 'clone' ]; then
				opts=$(get_vms)
			else
				opts=$(get_snaps_names)
			fi
			;;
		--location)
			COMPREPLY=($(compgen -o dirnames -- "${cur}"))
			return 0
			;;
		-c|--config|--file|--device|--image)
			# TODO Add filtering on file extentions
			COMPREPLY=($(compgen -A file -- "${cur}"))
			return 0
			;;
		--ostype)
			opts=$(get_ostypes)
			;;
		--distribution)
			opts=$(get_distros)
			;;
		--ostemplate)
			opts=$(get_ostemplates)
			;;
		--output)
			opts=$(get_vm_ids)
			;;
		-s|--sort)
			opts='name -name'
			;;
		--onboot)
			opts='yes no auto'
			;;
		--device-add)
			opts='hdd cdrom net fdd serial usb'
			;;
		--device-del)
			opts=$(get_devs "${COMP_WORDS[2]}")
			;;
		--device-connect)
			opts=$(get_devs "${COMP_WORDS[2]}" 'disconnected')
			;;
		--device-disconnect)
			opts=$(get_devs "${COMP_WORDS[2]}" 'connected')
			;;
		--device-set)
			opts=$(get_devs "${COMP_WORDS[2]}")
			;;
		--ioprio)
			# ioprio range: 0 - 7 (default : 4)
			opts='0 1 2 3 4 5 6 7'
			;;
		--cpus)
			# max vcpus
			local cores=$(grep '^cpu cores' /proc/cpuinfo | wc -l )
			local proc=$(grep 'processor' /proc/cpuinfo | wc -l)
			opts=$[ $cores * $proc ]
			;;
		--memsize)
			# max available memory
			opts=$(cat /proc/meminfo | grep  "MemTotal:" | awk '{print $2}')
			;;
		--autostart)
			opts="on off auto"
			;;
		--autostop)
			opts="stop suspend"
			;;
		--vnc-mode)
			opts="auto manual off"
			;;
		--ha-enable)
			opts='yes no'
			;;
		--path)
			COMPREPLY=($(compgen -o dirnames -- "${cur}"))
			return 0
			;;
		--type)
			# TODO for hdd: ide|scsi, for net:host|shared|bridget
			opts=''
			;;
		--tag)
			opts="$(get_backups_ids "${COMP_WORDS[2]}")"
			;;
		--id)
			local cmd=${COMP_WORDS[1]}
			case ${cmd} in
			snapshot-*)
				opts="$(get_snaps_ids "${COMP_WORDS[2]}")"
				[ -z "${cur}" ] && cur='{'
				;;
			restore)
				opts="$(get_backups_ids "${COMP_WORDS[2]}")"
				;;
			esac
			;;
		-o)
			if [ "${COMP_WORDS[1]}" = 'create' ]; then
				opts=$(get_ostypes)
			else
				opts=$(get_vm_ids)
			fi
			;;
		-i)
			action="${COMP_WORDS[1]}"
			if [ "${action::8}" = 'snapshot' ]; then
				# TODO See '--id'
				# opts=$(get_snaps_ids)
				opts=""
			elif [ "${action}" = 'list' ]; then
				opts="${list_flags} ${global_flags}"
			else
				opts=$(get_ostypes)
			fi
			;;
		--autostart)
			opts='stop suspend'
			;;
		--autostop)
			opts='on off auto'
			;;
		--applyconfig)
			COMPREPLY=($(compgen -A file -- "${cur}"))
			return 0
			;;
		--netfilter)
			opts='disabled stateless stateful full'
			;;
		--filter)
			# TODO find out names for statistics filering
			opts=''
			;;
		-l)
			# user:passwd@server
			opts=''
			;;
		-p|--read-passwd)
			COMPREPLY=($(compgen -A file -- "${cur}"))
			return 0
			;;
		-v)
			# TODO throw out -v
			opts=$global_flags
			;;
		--verbose)
			# TODO find out acceptable log levels
			opts='1 2 3'
			;;
		--vmtype)
			opts='ct vm all'
			;;

		*) # processing actions' local options

			case "${COMP_WORDS[1]}" in
			backup)
				opts="${backup_flags} ${global_flags}" 
				;;
			backup-delete)
				if [ $COMP_CWORD == 3 ]; then
					opts="${backupdelete_flags}" 
				else
					opts="${global_flags}"
				fi
				;;
			backup-list)
				opts="${backuplist_flags} ${global_flags}" 
				;;
			capture)
				[ $COMP_CWORD == 3 ] && opts="${capture_flags}"
				;;
			clone)
				case "${COMP_CWORD}" in
				3)
					opts="${clone_flags}"
					;;
				*)
					opts="${clone_optional_flags} ${global_flags}"
					;;
				esac
				;;
			create)
				# TODO Add exceptions for different parameter
				# configurations
				[ $COMP_CWORD == 3 ] && opts="${create_flags} ${global_flags}"
				;;
			delete)
				opts="${global_flags}"
				;;
			installtools)
				opts="${global_flags}"
				;;
			enter)
				opts="${global_flags}"
				;;
			list)
				opts="${list_flags} ${global_flags}"
				;;
			migrate)
				opts="${migrate_flags} ${global_flags}"
				;;
			move)
				case "${COMP_CWORD}" in
				3)
					opts="${move_flags}"
					;;
				*)
					opts="${global_flags}"
					;;
				esac
				;;
			pause)
				opts="${global_flags}"
				;;
			register)
				if [ $COMP_CWORD == 2 ]; then
					COMPREPLY=($(compgen -A file -- "${cur}"))
					return 0
				else
					opts="$global_flags"
				fi
				;;
			reset)
				opts="${global_flags}"
				;;
			reset-uptime)
				opts="${global_flags}"
				;;
			resume)
				opts="${global_flags}"
				;;
			restore)
				opts="${restore_flags}"
				;;
			server)
				if [ $COMP_CWORD == 2 ]; then
					opts='shutdown info'
				else
					opts="${global_flags}"
				fi
				;;
			start)
				opts="${global_flags}"
				;;
			stop)
				opts="${global_flags}"
				;;
			snapshot)
				opts="${snapshot_flags} ${global_flags}" 
				;;
			snapshot-delete)
				if [ $COMP_CWORD == 3 ]; then
					opts="${snapshotdelete_flags}" 
				else
					opts="${global_flags}"
				fi
				;;
			snapshot-list)
				opts="${snapshotlist_flags} ${global_flags}" 
				;;
			snapshot-switch)
				if [ $COMP_CWORD == 3 ]; then
					opts="${snapshotswitch_flags}" 
				else
					opts="${global_flags}"
				fi
				;;
			suspend)
				opts="${global_flags}"
				;;
			statistics)
				opts="${statistics_flags} ${global_flags}"
				;;
			unregister)
				opts="${global_flags}"
				;;
			set)
				opts="${set_flags} ${global_flags}"
				;;
			*)
				;;
			esac
			;;
		esac

	opts=$(for name in $opts; do printf "%q\n" "$name"; done)

	fi

	local OLDIFS=$IFS
	IFS=$'\n'

	COMPREPLY=($(compgen -W "${opts}" -- "${cur}"))

	IFS=$OLDIFS

	return 0
}

complete -o filenames -F _prlctl prlctl
